### PR TITLE
Limit the Number of Snapshots in a BlobStoreRepository

### PR DESCRIPTION
**Pull request by original-brownbear**
Motivated by a number of recent SDHs that resulted from incorrect SLM settings creating repositories that contained thousands of snapshots and eventually destabilised master nodes.

Adds a limit to the maximum number of snapshots that are allowed
to be added to a snapshot repository as a safety measure of last resort
against repositories that grow to an unmanagable size due to e.g. incorrect SLM
settings.

NOTE: This is just intended as a last ditch safety measure to avoid destabilising master nodes. It hooks into the last step of the snapshot state machine so a lot of redundant work is performed for snapshots that won't make it into the repository in the end. I chose this approach since it was significantly simpler to implement than the alternative of exposing a limit to the `SnapshotsService` which would then work out whether to start a snapshot based on the size of the repo and already running snapshots. The later would have required a number of changes which did not seem justified when to date all cases where this would have kicked in where the result of SLM issues (especially when the known ones are pratically fixed by https://github.com/elastic/elasticsearch/pull/62284 for now).

### CODE REVIEW

File: docs/reference/snapshot-restore/register-repository.asciidoc [hunk 1]
```diff
@@ -162,6 +162,9 @@ unit, for example: `1GB`, `10MB`, `5KB`, `500B`. Defaults to `null` (unlimited c
 `max_restore_bytes_per_sec`:: Throttles per node restore rate. Defaults to unlimited. Note that restores are also throttled through <<recovery,recovery settings>>.
 `max_snapshot_bytes_per_sec`:: Throttles per node snapshot rate. Defaults to `40mb` per second.
 `readonly`:: Makes repository read-only.  Defaults to `false`.
+`size_limit`:: Limits the maximum number of snapshots that the repository may contain. Defaults to `500`. Note that snapshot repositories do not
```
Comment (fcofdez) [inline docs/reference/snapshot-restore/register-repository.asciidoc:165]: This setting name might be confusing.  I would rename to `max_number_of_snapshots`.

**Review by fcofdez — COMMENTED** [review]
The changes LGTM but I'm wondering if this could be problematic for users that due to legal reasons have to keep a large number of snapshots. I'm not sure if those cases are common across Elasticsearch users. And I was thinking if we should make the default value unbounded and let users to configure that as they desire? That would reduce the friction during upgrades.

**Comment by original-brownbear** [comment]
> I'm not sure if those cases are common across Elasticsearch users. 

Hard to say how common they are, but they won't work past a certain number of snapshots. It's dependent on your hardware/heap-size/number_of_shards etc. but I would say once you get to 1k+ snapshots you're in trouble in most cases no matter what. If we set this to `500` by default I'd assume we would barely break any functioning setups (I have never seen anyone continuously operate a repo with 500+ snapshots in fact).

> And I was thinking if we should make the default value unbounded and let users to configure that as they desire? That would reduce the friction during upgrades.

The thing is, normally you'd try and avoid the situation by setting the correct SLM retention rules. This setting/limit I intended to be a fail-safe measure in case a mistake is made there or an unexpected bug occurs. Until today, we were able to fix all situations where the repo grew to an unmanageable size by creating a new repo or spending endless hours deleting old snapshots manually and then fixing the SLM settings. If we leave it unbounded by default we wouldn't have prevented any of these issues. Also, I think setting a limit here by default is in line with what we do in other parts of the code where we put soft+hard limits by default to avoid things blowing up.

File: docs/reference/snapshot-restore/register-repository.asciidoc (hunk 1 above)
Comment (original-brownbear) [reply docs/reference/snapshot-restore/register-repository.asciidoc:165]: ++

**Comment by original-brownbear** [comment]
I linked one SDH which is a good example of what this is trying to prevent :)

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/default-distro

**Comment by fcofdez** [comment]
>The thing is, normally you'd try and avoid the situation by setting the correct SLM retention rules. This setting/limit I intended to be a fail-safe measure in case a mistake is made there or an unexpected bug occurs. Until today, we were able to fix all situations where the repo grew to an unmanageable size by creating a new repo or spending endless hours deleting old snapshots manually and then fixing the SLM settings. If we leave it unbounded by default we wouldn't have prevented any of these issues. Also, I think setting a limit here by default is in line with what we do in other parts of the code where we put soft+hard limits by default to avoid things blowing up.

Yes, that seems reasonable 👍 . 

Thanks for the clarifications @original-brownbear!

**Review by fcofdez — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Francisco! 

@tlrx could you also take a look here just to make sure we're not missing any side-effects here potentially?

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/bwc
Jenkins run elasticsearch-ci/default-distro

(some composable template BwC issue)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 2]
```diff
@@ -1081,6 +1094,14 @@ public void finalizeSnapshot(final ShardGenerations shardGenerations,
         final StepListener<RepositoryData> repoDataListener = new StepListener<>();
         getRepositoryData(repoDataListener);
         repoDataListener.whenComplete(existingRepositoryData -> {
+            final int existingSnapshotCount = existingRepositoryData.getSnapshotIds().size();
+            if (existingSnapshotCount >= maxSnapshotCount) {
+                listener.onFailure(new RepositoryException(metadata.name(), "Can not add another snapshot to this repository as it " +
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1099]:
```suggestion
                listener.onFailure(new RepositoryException(metadata.name(), "Cannot add another snapshot to this repository as it " +
```

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java [hunk 3]
```diff
@@ -1081,6 +1094,14 @@ public void finalizeSnapshot(final ShardGenerations shardGenerations,
...
+                        "already contains [" + existingSnapshotCount + "] snapshots and is configured to hold up to [" + maxSnapshotCount +
+                        "] snapshots only. Please increase repository setting [" + MAX_SNAPSHOTS_SETTING.getKey() +
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:1101]: I think we should name the setting in the message (as done here) but think we shouldn't recommend increasing it if you hit the limit and should instead suggest deleting some snapshots and/or adjusting retention settings to bring the snapshot count below the limit.

File: docs/reference/snapshot-restore/register-repository.asciidoc [hunk 4]
```diff
@@ -162,6 +162,9 @@ unit, for example: `1GB`, `10MB`, `5KB`, `500B`. Defaults to `null` (unlimited c
 `max_restore_bytes_per_sec`:: Throttles per node restore rate. Defaults to unlimited. Note that restores are also throttled through <<recovery,recovery settings>>.
 `max_snapshot_bytes_per_sec`:: Throttles per node snapshot rate. Defaults to `40mb` per second.
 `readonly`:: Makes repository read-only.  Defaults to `false`.
+`max_number_of_snapshots`:: Limits the maximum number of snapshots that the repository may contain. Defaults to `500`. Note that snapshot repositories do not
+scale indefinitely in size and might lead to master node performance and stability issues if grown past a certain size. Increasing this setting
```
Comment (DaveCTurner) [inline docs/reference/snapshot-restore/register-repository.asciidoc:166]:
```suggestion
scale indefinitely in size and might lead to master node performance and stability issues if they grow past a certain size. We do not recommend increasing this setting.
```

File: docs/reference/snapshot-restore/register-repository.asciidoc [hunk 5]
```diff
@@ -162,6 +162,9 @@ unit, for example: `1GB`, `10MB`, `5KB`, `500B`. Defaults to `null` (unlimited c
...
+is generally not advisable and should be weighted carefully against alternatives such as deleting older snapshots or creating multiple repositories.
```
Comment (DaveCTurner) [inline docs/reference/snapshot-restore/register-repository.asciidoc:167]:
```suggestion
Instead you should delete older snapshots or use multiple repositories.
```

Is creating multiple repositories really a solution?

**Review by DaveCTurner — COMMENTED** [review]
Some wording nits/suggestions.

File: docs/reference/snapshot-restore/register-repository.asciidoc (hunk 5 above)
Comment (original-brownbear) [reply docs/reference/snapshot-restore/register-repository.asciidoc:167]: I guess technically yes, not a great one but you could run two (or more) SLM policies that take turns creating snapshots I suppose. That way you could have a much longer retention time at high resolution. 
Also, if you just want certain snapshots to live for a long time maybe it's a solution.
... certainly all of these aren't super ergonomically but I guess for some use-cases these are valid?

**Review by tlrx — APPROVED** [review]
LGTM. I agree with David's suggestion about wording. I'm a bit afraid that it would cause trouble to existing large repositories so I suggest that it goes to 8.0 only and unbounded in 7.x with log warnings.

**Comment by original-brownbear** [comment]
>  I'm a bit afraid that it would cause trouble to existing large repositories 

I'm not to be honest :) We log warnings already (due to the size of the `RepositoryData` in this case). This is really just about preventing breakage in your repository. Even 500 snapshots is quite a number with SLM. This means that SLM will sequentially have to load all `SnapshotInfo` for those 500 snapshots which takes multiple minutes in many cases and is a completely unstable situation. If we actually break a repo with this, it's trivial to fix for the user by upping the setting and in many cases we had with this many snapshots in a repo, people just adjusted their retention parameters since they weren't aware of the issue. 

-> I'd rather merge this into 7.x to prevent issues as early as possible?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java]: ++ adjusted the exception message

**Comment by tlrx** [comment]
@original-brownbear I understand your reasoning but I think we usually introduce such limits in major versions only and emit deprecation warnings in the next minor along with documenting the change. I don't see why we would make an exception here?

**Comment by original-brownbear** [comment]
Thanks everybody, I'm merging this to `master` and as discussed with Tanguy in person will leave the if and how of a a backport to after we had a chance to discuss the impact of this on 7.x in the snapshot resiliency meeting.

**Pull request merged**